www/utils/supabase/queries.ts

```diff
@@ -1,5 +1,42 @@
 import { cache } from 'react';
 import { SupabaseClient } from '@supabase/supabase-js';
+import { unstable_cache as next_unstable_cache } from 'next/cache';
+import { getFreeMessageCount } from './actions';
+
+export const unstable_cache = <Args extends any[], Output>(
+  callback: (...args: Args) => Promise<Output>,
+  key: string[],
+  options: { revalidate: number },
+) => {
+  return cache(
+    next_unstable_cache(
+      callback as unknown as (...args: any[]) => Promise<Output>,
+      key,
+      options
+    )
+  );
+};
+
+export const getChatAccess = unstable_cache(
+  async (supabase: SupabaseClient, userId: string) => {
+    const { data: subscription } = await supabase
+      .from('subscriptions')
+      .select('status, trial_end')
+      .in('status', ['trialing', 'active'])
+      .maybeSingle();
+    
+    const isSubscribed = !!(subscription && subscription.status === 'active' && !subscription.trial_end);
+    const freeMessages = isSubscribed ? 0 : await getFreeMessageCount(userId);
```

Comment:
Once you have the subscription already why do you need to do another call to getFreeMessageCount? This should already be in the metadata. You're repeating the same query multiple times here.